kubeadm: enhanced TLS validation for token-based discovery in `kubeadm join`

mattmoyer opened the pull request:
**What this PR does / why we need it**:
This PR implements enhanced TLS validation for `kubeadm join` when using token-based TLS discovery. Without this enhancement, `kubeadm join` has some less-than-ideal security properties. Specifically, in the case where a bootstrap token is compromised, the attacker can impersonate the API server to newly bootstrapping clients ([more discussion in the design proposal](https://docs.google.com/document/d/1SP4P7LJWSA8vUXj27UvKdVEdhpo5Fp0QHNo4TXvLQbw/edit?ts=5971498a)).

The gist of this enhancement is to support public key pinning in the style of [RFC7469](https://tools.ietf.org/html/rfc7469#section-2.4). When bootstrapping, `kubeadm` can now be configured with a whitelist of root CA public keys. It can then validate that the cluster it connects to is operated by the owner of one of those public keys.

These public key hashes are short enough that the entire `kubeadm join` command can still be copy-pasted relatively easily (not as easily as before, but ~160 characters). Using a public key hash rather than a hash over the entire certificate allows certificates to be reissued with updated expirations without invalidating existing key pins.

This change adds two new command line flags (and associated config parameters):
 - **`--discovery-token-ca-cert-hash sha256:<hash>`:**
   Validates that the cluster root CA has a public key fingerprint that matches one of the specified values. If this flag is not passed when token-based discovery is being used, a warning is printed. This warning will become an error in 1.9.
 - **`--discovery-token-unsafe-skip-ca-verification`:**
    Disables the warning message when no keys are pinned. In 1.9, this flag will be required _unless_ `--discovery-token-unsafe-skip-ca-verification` is used.

This is fully backwards compatible and client side (kubeadm) only. It will be a breaking change when the flag becomes required in v1.9.

This validation is done after and in addition to the existing bootstrap token signing/MAC mechanism.

#### Example from `kubeadm init`:
```
$ kubeadm init
[...]

You can now join any number of machines by running the following on each node
as root:

  kubeadm join --token a66ae0.1f8a5ed9a210e187 192.168.42.10:6443 --discovery-token-ca-cert-hash sha256:547c102383c0f26387b961b4e9b8f842dc07c074c8316f238dbcf5563fc3ac35
```

#### Example from `kubeadm join`:
```
$ kubeadm join --token a66ae0.1f8a5ed9a210e187 192.168.42.10:6443 --discovery-token-ca-cert-hash sha256:547c102383c0f26387b961b4e9b8f842dc07c074c8316f238dbcf5563fc3ac35
[kubeadm] WARNING: kubeadm is in beta, please do not use it for production clusters.
[preflight] Running pre-flight checks
[discovery] Trying to connect to API Server "192.168.42.10:6443"
[discovery] Created cluster-info discovery client, requesting info from "https://192.168.42.10:6443"
[discovery] Requesting info from "https://192.168.42.10:6443" again to validate TLS against the pinned public key
[discovery] Cluster info signature and contents are valid and TLS certificate validates against pinned roots, will use API Server "192.168.42.10:6443"
[discovery] Successfully established connection with API Server "192.168.42.10:6443"

Node join complete:
* Certificate signing request sent to master and response
  received.
* Kubelet informed of new secure connection details.

Run 'kubectl get nodes' on the master to see this machine join.
```

**Which issue this PR fixes**:
ref https://github.com/kubernetes/features/issues/130
fixes: https://github.com/kubernetes/kubeadm/issues/365

**Special notes for your reviewer**:
This was proposed and discussed briefly by SIG-cluster-lifecycle and SIG-auth. The design proposal is [in Google Docs](https://docs.google.com/document/d/1SP4P7LJWSA8vUXj27UvKdVEdhpo5Fp0QHNo4TXvLQbw/edit?ts=5971498a).

There is a documentation change needed to explain the security properties of `kubeadm join` with and without `--discovery-token-ca-cert-hash`. This page should be linked by to by the warning message when you don't pass either of the new flags (I have it pointing [here](https://kubernetes.io/docs/admin/kubeadm/#kubeadm-join) for now, which I think will be the right place). I will follow up with this documentation shortly.

**Release note**:
```release-note
kubeadm: added enhanced TLS validation for token-based discovery in `kubeadm join` using a new `--discovery-token-ca-cert-hash` flag.
```

/cc @luxas @jbeda @ericchiang

mattmoyer committed:
kubeadm: implement TLS discovery root CA pinning.

This change adds the `--discovery-token-ca-cert-hash` and `--discovery-token-unsafe-skip-ca-verification` flags for `kubeadm join` and corresponding fields on the kubeadm NodeConfiguration struct. These flags configure enhanced TLS validation for token-based discovery.

The enhanced TLS validation works by pinning the public key hashes of the cluster CA. This is done by connecting to the `cluster-info` endpoint initially using an unvalidated/unsafe TLS connection. After the cluster info has been loaded, parsed, and validated with the existing symmetric signature/MAC scheme, the root CA is validated against the pinned public key set. A second request is made using validated/safe TLS using the newly-known CA and the result is validated to make sure the same `cluster-info` was returned from both requests.

This validation prevents a class of attacks where a leaked bootstrap token (such as from a compromised worker node) allows an attacker to impersonate the API server.

This change also update `kubeadm init` to print the correct `--discovery-token-ca-cert-hash` flag in the example `kubeadm join` command it prints at the end of initialization.

mattmoyer committed:
kubeadm: generated deepcopy for `k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm` and `k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1alpha1`.

mattmoyer committed:
kubeadm: add pubkeypin package (public key pinning hash implementation).

This change adds a `k8s.io/kubernetes/cmd/kubeadm/app/util/pubkeypin` package which implements x509 public key pinning in the style of RFC7469. This is the public key hash format used by the new `kubeadm join --discovery-token-ca-cert-hash` flag.

Hashes are namespaced with a short type, with "sha256" being the only currently-supported format. Type "sha256" is a hex-encoded SHA-256 hash over the Subject Public Key Info (SPKI) object in DER-encoded ASN.1.

ericchiang commented:
/ok-to-test

High level opinion that I brought up in the design doc, but wanted to reiterate here.

The value of `--tls-discovery-root-ca` for `kubeadm join` can't be known until `kubeadm init` runs on the master to generate the CA. This means I can't provision my workers until my masters have been provisioned and `kubeadm init` has run on the master

This won't work for anyone who provisions the masters and the workers at the same time. For example, _Kubernetes Anywhere_ would have to use the `--tls-discovery-allow-unsafe`:

https://github.com/kubernetes/kubernetes-anywhere/blob/8d2c7788c63693b0021278dedcf03fad935efef1/phase1/gce/configure-vm-kubeadm.sh#L40

I am for this fix, but I think a majority of auto-scaling setups will have to use the unsafe flag.

ericchiang commented on cmd/kubeadm/app/discovery/token/token.go:
```diff
@@ -131,3 +205,76 @@ func runForEndpointsAndReturnFirst(endpoints []string, fetchKubeConfigFunc func(
+	}
+
+	// Perform what should be a single request using the client
+	response, err := doRequest(&insecureClient)
```
Need to close the response

ericchiang commented on cmd/kubeadm/app/discovery/token/token.go:
```diff
@@ -131,3 +205,76 @@ func runForEndpointsAndReturnFirst(endpoints []string, fetchKubeConfigFunc func(
+	var certificates certChain
+
+	// Create an HTTP client with a custom TLS transport that saves off the certificate chain for later validation
+	insecureClient := http.Client{
```
We had issues constantly re-initializing transports like this since they have a tendency to keep TCP connections alive.

mattmoyer replied:
That makes sense. I think it shouldn't be a huge issue for `kubeadm` since it's relatively short-lived. That's still pretty non-ideal, though.

I'll see if I can find a way to rework this so the http.Client is long lived.

ericchiang commented on cmd/kubeadm/app/discovery/token/token.go:
```diff
@@ -131,3 +205,76 @@ func runForEndpointsAndReturnFirst(endpoints []string, fetchKubeConfigFunc func(
+
+	// Make sure the wrapper is being used correctly
+	if len(certificates) == 0 {
+		return nil, nil, errInsecureClientNoRequest
```
If you error here you won't close the response body.

mattmoyer replied:
Ah, thanks. Will fix.

ericchiang commented on cmd/kubeadm/app/discovery/token/token.go:
```diff
@@ -27,44 +34,81 @@ import (
 
 const BootstrapUser = "token-bootstrap-client"
 
+var (
```
Does kubeadm normally declare errors like this? I'd expect the fmt.Errorf call to just be inlined. e.g. https://github.com/kubernetes/kubernetes/blob/93b144c426385b3680cb608ba882141a1ec85841/cmd/kubeadm/app/util/config/masterconfig.go#L58

mattmoyer replied:
Yeah, this was mainly so I can test that particular error cases return the correct error.

luxas replied:
I'm ok with it, but we aren't doing it in other places. This seems to be a good place to use this method though...

ericchiang commented on cmd/kubeadm/app/discovery/token/token.go:
```diff
@@ -27,44 +34,81 @@ import (
+			defer response.Body.Close()
+
+			// Read the entire response body
+			responseBytes, err := ioutil.ReadAll(response.Body)
```
You don't trust the server at this point. Probably good practice to limit the number of bytes read to something reasonable.

mattmoyer replied:
+1 will fix.

ericchiang commented on cmd/kubeadm/app/discovery/token/token.go:
```diff
@@ -85,22 +129,52 @@ func RetrieveValidatedClusterInfo(discoveryToken string, tokenAPIServers []strin
+			chainCerts.AddCert(chainCert)
+		}
+
+		// Now that we know the root CA pool, validate the original certificate chain
```
It seems much simpler and less prone to error to just make a second request to get the cluster-info once you have the CA and let the Go TLS library automatically do the validation.

mattmoyer replied:
Making double requests seemed worse to me, but I think it's sound from a security perspective.

luxas commented on cmd/kubeadm/app/discovery/token/token.go:
```diff
@@ -131,3 +205,76 @@ func runForEndpointsAndReturnFirst(endpoints []string, fetchKubeConfigFunc func(
+// which should perform a single HTTP request and return the response. The temporary
+// client will not validate TLS certificates in real time, but instead will capture
+// and return the certificates for post-hoc validation.
+func withInsecureHTTPClient(
```
Is there no way to save the cert chain with the normal kube client?
So far we've been able to just skip TLS verification with the normal kube client.

mattmoyer replied:
We could extend the normal API client to give more control over the underlying `http.Client`. I think extending it to support this particular case wouldn't be worth it, but a more general change that allowed passing in a custom `http.Client` could be more generally useful.

The thinking with doing this as a "raw" `http.Client` instead of using the normal API client was that this is fairly simple from an HTTP standpoint (single GET request) and that the TLS bits were such an edge case that we probably don't want that complexity in the main API client code.

luxas commented on cmd/kubeadm/app/util/pubkeypin/pubkeypin.go:
```diff
@@ -0,0 +1,77 @@
+
+// Package pubkeypin provides primitives for x509 public key pinning in the
+// style of RFC7469.
+package pubkeypin
```
Do we consider this kubeadm-specific or can we move this to `pkg/util` or something like `k8s.io/apimachinery/pkg/util/sets`?

mattmoyer replied:
Right now `kubeadm` is the only thing that knows or cares about these public key pins. It might be useful somewhere else (maybe in `kubectl` also?).

I don't feel strongly about this either way.

ericchiang replied:
Please keep it in `kubeadm`. The more dependencies we create, the harder it's going to be to split the repos in the future.

luxas commented:
> [discovery] Failed to connect to API Server "192.168.42.10:6443": there is no JWS signed token in the cluster-info ConfigMap. This token id "ec2651" is invalid for this cluster, can't connect

@mattmoyer Are you sure why this happened? Was the token actually not there or is it an error in this code?

> The value of --tls-discovery-root-ca for kubeadm join can't be known until kubeadm init runs on the master to generate the CA. This means I can't provision my workers until my masters have been provisioned and kubeadm init has run on the master

@ericchiang The user still has the power to pre-provision the CA, so they can know what the CA hash will be in beforehand
Basically doing this:
```
// generate CA private key
openssl --foo -out /etc/kubernetes/pki/ca.key
// generate CA certificate
openssl --bar -out /etc/kubernetes/pki/ca.crt
// get CA hash
openssl x509 -pubkey -in /etc/kubernetes/pki/ca.crt | openssl rsa -pubin -outform der 2>&/dev/null | openssl dgst -sha256 -hex > my-CA-hash.txt

kubeadm init --token <id>.<secret>
```

mattmoyer commented on cmd/kubeadm/app/util/pubkeypin/pubkeypin.go:
```diff
@@ -0,0 +1,77 @@
+
+// Set is a set of pinned x509 public keys.
+type Set struct {
+	sha256Hashes [][]byte
```
+1, will fix. I originally had this code supporting other hash types/lengths (truncated SHA-256/128 and SHA-256 encoded as base58). In the end I just left plain hex-encoded SHA-256 so this doesn't make a lot of sense.

mattmoyer commented on cmd/kubeadm/app/discovery/token/token.go:
```diff
@@ -131,3 +205,76 @@ func runForEndpointsAndReturnFirst(endpoints []string, fetchKubeConfigFunc func(
+type certChain []*x509.Certificate
+
+// saveCertificateChain parses and saves the array of DER-encoded certificates into a certChain
+func (c *certChain) saveCertificateChain(rawCerts [][]byte, _ [][]*x509.Certificate) error {
```
Yeah, validating the entire initial connection gives better trust in the other parts of the `cluster-info` ConfigMap since we use it for more than just the cluster CA. I think we could also accomplish this by making two requests if we want to go that route.

We need the full chain since the server might have some intermediate CAs included that we need to chain through to get from the root to the leaf cert.

ericchiang commented:
> The user still has the power to pre-provision the CA, so they can know what the CA hash will be in beforehand

@luxas ah, great! Is this actual endorsed behavior or a nice consequence of the code? Might be worth updating the docs as part of this PR.

mattmoyer commented on cmd/kubeadm/app/apis/kubeadm/validation/validation.go:
```diff
@@ -135,6 +135,17 @@ func ValidateArgSelection(cfg *kubeadm.NodeConfiguration, fldPath *field.Path) f
+	}
+
+	// TODO: convert this warning to an error after v1.8
+	if len(cfg.DiscoveryFile) == 0 && len(cfg.TLSDiscoveryRootCAPubKeys) == 0 && !cfg.TLSDiscoveryAllowUnsafe {
```
The way it's written now, `--tls-discovery-root-ca` _only_ applies to the plain token-based case. If you're using `DiscoveryFile` with local files or a remote HTTPS endpoint, it doesn't do any validation. 

I think that's reasonable because `DiscoveryFile` should be transmitted through some secure means that protects its integrity. Either the local file will have been copied there by some secure orchestrator or the remote file will be authenticated by the web PKI and the system roots. I don't think we need to consider these cases unsafe or require the user to pass `--tls-discovery-allow-unsafe`.

I have the validation error `TLSDiscoveryRootCAPubKeys cannot be used with DiscoveryFile` to make sure you don't get a false sense of security from thinking you've pinned the CA when you're actually just relying on the security of the discovery file.

mattmoyer commented on cmd/kubeadm/app/discovery/token/token.go:
```diff
@@ -27,44 +34,81 @@ import (
+			bootstrapapi.ConfigMapClusterInfo))
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("invalid URL, can't connect: %v", err)
```
Thanks, will fix.

ericchiang commented:
cc @kubernetes/sig-auth-pr-reviews

mattmoyer commented:
@luxas 
> > [discovery] Failed to connect to API Server "192.168.42.10:6443": there is no JWS signed token in the cluster-info ConfigMap. This token id "ec2651" is invalid for this cluster, can't connect

 > @mattmoyer Are you sure why this happened? Was the token actually not there or is it an error in this code?

I think this was just an artifact of my local vagrant environment. It's not reproducible locally, so I think the server just wasn't up quite yet in this case.

jbeda commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -101,4 +101,16 @@ type NodeConfiguration struct {
 	Token                    string
+
+	// TLSDiscoveryRootCAPubKeys specifies a set of public key pins. The root CA
+	// discovered during TLS bootstrapping must match one of these values.
```
Technically, this is during the bootstrap discovery, not TLS.  We use this so that we can establish the kubelet trusting the master.

mattmoyer replied:
Ah, yeah I'll fix that comment and see if I made the same mistake elsewhere. I originally approached this code without a clear understanding of the "discovery" vs. "TLS bootstrapping" terminology.

jbeda commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -101,4 +101,16 @@ type NodeConfiguration struct {
+
+	// TLSDiscoveryRootCAPubKeys specifies a set of public key pins. The root CA
+	// discovered during TLS bootstrapping must match one of these values.
+	// Specifying an empty set disables root CA pinning. Each hash is a hex encoded
```
I'd still love to see us do something like base 36 or base 58 here to shorten this.  But that can come in a later PR.

mattmoyer replied:
I tried to implement the `pubkeypin` so it's easy to swap in support for multiple encodings and lengths.

I think if we do that we should also switch the default token encodings as well.

jbeda reviewed (approved):
LGTM.

I'm torn on the "validate current conn" vs. "make a new conn".  I could go either way.  So I won't express an opinion.

mattmoyer commented:
`pull-kubernetes-kubemark-e2e-gce` failure is https://github.com/kubernetes/kubernetes/issues/49320

mattmoyer commented on cmd/kubeadm/app/discovery/token/token.go:
```diff
@@ -85,22 +129,52 @@ func RetrieveValidatedClusterInfo(discoveryToken string, tokenAPIServers []strin
+
+		// For each cluster, validate that its CA matches a pinned key and save that root into an x509.CertPool
+		pinnedRoots := x509.NewCertPool()
+		for _, cluster := range finalConfig.Clusters {
```
Thanks. Instead of looping I added an assertion that there is only a single cluster.

timothysc reviewed (commented):
going to recuse myself from this one other then to say, please squash commits.

mattmoyer commented:
I think the latest change addresses all the reviews so far, including ripping out the post-hoc validation in favor of just making two requests.

I'm leaving the original commits in https://github.com/mattmoyer/kubernetes/commits/8d9feb79c1619d8a80350f8ce1aa6dbd47350e89 and will squash/rebase this PR against current `master` for a final review.

Thanks for all the feedback.

mattmoyer commented:
/retest

mattmoyer commented:
/retest

mattmoyer commented:
/retest

mattmoyer commented:
/retest

mattmoyer commented:
Failing test looks to be https://github.com/kubernetes/kubernetes/issues/49672

mattmoyer commented:
/retest

jbeda commented:
Let's start the lazy consensus clock ticking on this.  If we don't have any further concerns from EOD tuesday, I'll mark this LGTM on Wed.

luxas commented on cmd/kubeadm/app/discovery/token/token.go:
```diff
@@ -35,58 +39,123 @@ import (
-
+// It then makes sure it can trust the API Server by looking at the JWS-signed tokens and (if rootCAPubKeys is not empty)
+// validating the cluster CA against a set of pinned public keys
+func RetrieveValidatedClusterInfo(discoveryToken string, tokenAPIServers, rootCAPubKeys []string) (*clientcmdapi.Cluster, error) {
```
My only high-level comment here is that this function becomes a little bit too long in my opinion, but that's nothing we should hold up this PR for

mattmoyer replied:
Agreed. It's also lacking test coverage.

luxas replied:
Exactly, can't be unit tested currently.
However, I'm fine with deferring that to a follow-up, this PR is already quite big

luxas reviewed (approved):
Thanks! Now looks cleaner to me when you're reusing the built-in kube client to do this stuff instead of creating custom clients :+1:

/lgtm

luxas commented:
Only nit is as @timothysc said; squash your commits.
For instance:
1. Add generic util package for pubkeypins
2. Add the new actual code
3. Autogenerated

mattmoyer commented:
@luxas rebased from five commits down to the three you suggested.

Thanks again for the review.

mattmoyer commented:
/retest

mattmoyer commented:
Flaking `pull-kubernetes-bazel` looks like https://github.com/kubernetes/kubernetes/issues/49844

mattmoyer commented:
/test pull-kubernetes-federation-e2e-gce

luxas commented:
/retest

mattmoyer commented:
/retest

mattmoyer commented:
/retest

ericchiang reviewed (approved):
Thanks for the change.

/lgtm

luxas commented:
/lgtm

luxas commented:
I now created https://github.com/kubernetes/kubeadm/issues/365 and added a "fixes" link in the description, so the bot would be able to pick it up.
`hack/verify-flags` is an approving mess, currently discussing with sig-testing how we can get rid of it

mikedanese commented:
Just saw this when lucas poked on slack. I'm going to give this a review now. I'm mostly concerned about the commandline flag names. Also it would be helpful to move the PR description into an update of the discovery design proposal.

mikedanese commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -101,4 +101,16 @@ type NodeConfiguration struct {
+	// This is similar to the format defined in https://tools.ietf.org/html/rfc7469#section-2.4.
+	// These hashes can be calculated using, e.g.,
+	// `openssl x509 -pubkey -in ca.crt openssl rsa -pubin -outform der 2>&/dev/null | openssl dgst -sha256 -hex`
+	TLSDiscoveryRootCAPubKeys []string
```
Some problems with this:
* There is no indication that this is supposed to be a hash. Preferably the name should carry this indication.
* There is no indication that this is only relevant for token discovery. Preferably the name should carry this indication.
* We should namespace hashes so we can support multiple hashes incase we need to move away from sha256 for whatever reason.
* There is no indication of the format of the hash serialization. Does hex work? Does base64 work? It's ok to note this in the field documentation.

Strawman:
* Change name to `DiscoveryTokenCACertHashes`
* Expect and document a serialization format of `<hash-type>:<base16-encoded-sum>` and a list of supported hashes starting with `sha256` only

mikedanese commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -101,4 +101,16 @@ type NodeConfiguration struct {
+	TLSDiscoveryRootCAPubKeys []string
+
+	// TLSDiscoveryAllowUnsafe disables warnings when token-based TLS discovery is used and TLSDiscoveryRootCAPubKeys is empty
+	TLSDiscoveryAllowUnsafe bool
```
I would also prefer if this was prefixed in a way that indicated it was only relevant for token discovery.

mikedanese replied:
perhaps `DiscoveryTokenUnsafeSkipCAVerification`

mikedanese commented on cmd/kubeadm/app/cmd/join.go:
```diff
@@ -116,6 +116,13 @@ func NewCmdJoin(out io.Writer) *cobra.Command {
 		"A token used for TLS bootstrapping")
+	cmd.PersistentFlags().StringSliceVar(
+		&cfg.TLSDiscoveryRootCAPubKeys, "tls-discovery-root-ca", []string{},
+		"Validate that the root CA public key matches this SHA-256 SPKI hash (repeat to allow multiple public keys).")
```
Flag doc and name should indicate that this is only relevant for token discovery.

mikedanese commented:
Why are we allowing unsafe at all after deprecation? Why not just remove the `TLSDiscoveryAllowUnsafe` instead of switiching it's default. Both changes seem equally disruptive.

jbeda commented:
I think that @mikedanese's comments are good around naming.  Making this clear it is a discovery thing is good.  Also like the idea of being more explicit in the hashing (I'd like to allow either base36 or base58 too, but that can come in another change).

In terms of "why allow unsafe at all", we are allowing folks to relax this for a "self stitching" cluster bring up where the cert hash isn't known at node bring up time.  That is covered in the design proposal.

mattmoyer commented:
@mikedanese : Thanks for the review. I agree that the names weren't perfect. I just updated with the names you suggested, added an explicit type on the hash value (so it's `sha256:<value>`), and expanded the usage/docs a little.

I'll update the PR body with the new flags as well.

mikedanese commented:
/lgtm

mikedanese commented:
/retest
